### PR TITLE
Тимакин - Лабораторная работа #2

### PR DESCRIPTION
**Pull request by owerbat**
@VadimKutovoi @pavjelv @TheG1uy
Please review

### CODE REVIEW

File: modules/red-black-tree/test/test_rb_tree.cpp [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+TEST(RedBlackTreeTest, Can_Get_Current) {
+    Node *node = new Node(1);
+    RBTree tree{node};
+
+    tree.begin();
+
+    ASSERT_EQ(1, tree.get_current()->value);
+}
+
+TEST(RedBlackTreeTest, Can_Find_Excisting_Item) {
```
Comment (pavjelv) [inline modules/red-black-tree/test/test_rb_tree.cpp]: Probably a typo if you meant Existing

**Comment by alexbatashev** [comment]
@owerbat не все методы покрыты тестами

**Comment by alexbatashev** [comment]
@overbat here is the full codecov report https://codecov.io/gh/UNN-ITMM-Software/devtools-course-practice/src/f0ffac1f4ef577aea7cc1206b4db0b21d9cd94c0/modules/red-black-tree/src/rb_tree.cpp

Make sure all branch cases fire at least once

**Comment by owerbat** [comment]
@alexbatashev Yes, I see. I just can't predict all branches the algorithm will visit. So I'll try to catch them all

File: modules/red-black-tree/include/rb_tree.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    explicit RBTree(Node *node);
+    explicit RBTree(const std::vector<int>& vec);
+
+    RBTree(const RBTree& tree);  // Don't use! It added because of syntax errors
+
+    Node* get_root() const;
+    Node* get_current() const;
+    unsigned get_nodes_number() const;
+
+    bool find(const int value);
```
Comment (alexbatashev) [inline modules/red-black-tree/include/rb_tree.h]: CamelCase for public methods

File: modules/red-black-tree/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Declare variables for binaries' names
+get_filename_component(DIR_NAME ${CMAKE_CURRENT_LIST_DIR} NAME)
+set(MODULE      "${DIR_NAME}")
+set(LIBRARY     "lib_${MODULE}")
+set(TESTS       "test_${MODULE}")
+set(APPLICATION "app_${MODULE}")
+
+# Include directory with public headers
+include_directories("./")
```
Comment (alexbatashev) [inline modules/red-black-tree/CMakeLists.txt]: CMAKE_CURRENT_SOURCE_DIR

File: modules/red-black-tree/test/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,27 @@
+set(target ${TESTS})
+
+file(GLOB srcs "*.cpp")
+set_source_files_properties(${srcs} PROPERTIES
+    LABELS "${MODULE};Test")
+
+# VS2012 doesn't support correctly the tuples yet,
```
Comment (alexbatashev) [inline modules/red-black-tree/test/CMakeLists.txt]: Remove if not used

File: modules/red-black-tree/test/test_rb_tree.cpp [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        vec.emplace_back(i);
+    std::random_shuffle(vec.begin(), vec.end());
+
+    RBTree tree{vec};
+
+    auto some_tree_operations = [&tree, &vec] {
+        for (tree.begin(); !tree.end(); tree.next())
+            Node *node = tree.get_current();
+
+        std::random_shuffle(vec.begin(), vec.end());
```
Comment (alexbatashev) [inline modules/red-black-tree/test/test_rb_tree.cpp]: Random data is not a good idea for a unit test. It may cause sporadic test results.

**Pull request merged**